#354 | track publishing failures

slayful opened the pull request:
## Proposed Changes

References #354, followed up with #360.

## Types of Changes

- [ ] Bugfix (non-breaking change which references issue #354)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)

## Checklist

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] All tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in related repositories

pivotal-issuemaster commented:
@slayful Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/rabbitmq?repositoryId=rabbitmq/rabbitmq-java-client&pullRequestId=357)!

[Click here](https://cla.pivotal.io/sync/rabbitmq?repositoryId=rabbitmq/rabbitmq-java-client&pullRequestId=357) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@slayful Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/rabbitmq?repositoryId=rabbitmq/rabbitmq-java-client&pullRequestId=357)!

acogoluegnes commented on src/main/java/com/rabbitmq/client/MetricsCollector.java:
```diff
@@ -36,6 +36,8 @@
 
     void basicPublish(Channel channel);
 
+    void basicPublishFailure(Channel channel);
```
I'd add the thrown exception as a second parameter. Our implementations wouldn't use it, but external ones could.

acogoluegnes commented:
Thanks for this contribution @slayful! I reviewed the PR, it doesn't need much change. Note this is a breaking change as it adds a method to an existing interface (I changed the PR summary to reflect this).

slayful commented:
Hey, @acogoluegnes I've addressed the feedback, so if you could take a look that would be great.

I can see that concourse-ci's not passing though I don't have access there. I'm currently running the tests on my localhost, and they're looking good so far.

slayful commented:
Ah, you already did, thank you! :)

acogoluegnes commented:
@slayful The Concourse failure was something else actually, but it's fixed now. Just waiting a bit more before merging, in case @michaelklishin comes up with something to change.

michaelklishin commented:
I at first expected this to include what's apparently in #360. That's fine. Perhaps we should add some docs about what exactly counts as a failure at the moment, since unroutable messages and negative acks will be separate metrics.
